interpreter/src/lib.rs

```diff
@@ -93,21 +168,110 @@ fn init(statements: &[Init], state: &mut State) {
 }
 
 fn run_threads(threads: &[Thread], state: &mut State) {
-    todo!()
+    let mut rng = rand::thread_rng(); // Create a random number generator
+    let mut active_threads = (0..threads.len()).collect::<Vec<_>>(); // Track active threads by their indices
+    let mut ip = vec![0; threads.len()]; // Instruction pointers for each thread
+    
+    for thread in threads {
+        state.write_buffers.entry(thread.name.clone()).or_insert_with(HashMap::new);
+    }
+
+    while !active_threads.is_empty() {
+        // Randomly select an active thread
+        let idx = rng.gen_range(0..active_threads.len());
+        let thread_idx = active_threads[idx];
+
+        // Run the next instruction if there is one
+        if ip[thread_idx] < threads[thread_idx].instructions.len() {
+            let instruction = &threads[thread_idx].instructions[ip[thread_idx]];
+            simulate_instruction(instruction, &threads[thread_idx].name, state);
+            ip[thread_idx] += 1; // Move the instruction pointer forward
+
+            // Check if this thread has completed all its instructions
+            if ip[thread_idx] >= threads[thread_idx].instructions.len() {
+                state.flush_write_buffer(&threads[thread_idx].name);
```

Comment:
I don't think the write buffer should be immediately flushed after the thread finishes since it prevents for some inter-leavings to occur